app/dao/notifications_dao.py

```diff
@@ -205,6 +205,10 @@ def country_records_delivery(phone_prefix):
 
 
 def _update_notification_status(notification, status, provider_response=None, bounce_response=None):
+    # Sometimes we get callback from the providers in the wrong order. If the notification has a
+    # permanent failure status, we don't want to overwrite it with a delivered status.
+    if notification.status == NOTIFICATION_PERMANENT_FAILURE and status == NOTIFICATION_DELIVERED:
```

Comment:
Although possibly benign, adding the check here changes the behaviour of several upstream methods, including some SMS stuff:
- `app/celery/process_ses_receipts_tasks.py`: `process_ses_results()`
- `app/celery/process_sns_receipts_tasks.py`: `process_sns_results()`
- `app/dao/notifications_dao.py`: `update_notification_status_by_id()`
- `app/dao/notifications_dao.py`: `update_notification_status_by_reference()`
- `app/notifications/notifications_ses_callback.py`: `handle_complaint()`
- `app/service/send_notification.py`: `send_one_off_notification()`

Alternatively we could do it specifically when we get the delivery receipt notification in `process_ses_receipts_tasks.py`.  

What do you think?

Reply:
changed!